Finetune lm parameters by fitting parabola curve

mychiux413 opened the pull request:
This is an option for tuning language model parameters: `alpha`, `beta` by fitting parabola curve , then report a json file to compare the different wer between: `oringal lm params` v.s. `finetuned lm params`

Due to time-consuming issue, this tool doesn't use grid search to iterate the whole datasets, but use `parabola fitting` ([paddle-deepspeech-grid-search](https://github.com/PaddlePaddle/DeepSpeech/blob/develop/docs/images/tuning_error_surface.png)) and just iterate `sampling data` to fast get good performance.

Fitting Process Step by Step:
  * samping the specified csv file: `--finetune_lm_csv_files` with sampling size = ` --finetune_lm_sampling_size` 
  * scan `alpha` linespace from `--finetune_lm_alpha_min` to `--finetune_lm_alpha_min` with steps `--finetune_lm_alpha_steps` (set `beta` = `--lm_alpha`)
  * fitting valley parabola curve  on `wer v.s. alpha`, finding `best alpha`(the bottom of valley), if the fitting is not an idea valley parabola, just pick the lowest wer from scanning points.
  * scan `beta` linespace from `--finetune_lm_beta_min` to `--finetune_lm_beta_min` (`alpha` = `best alpha`)
  * fitting valley parabola curve  on `wer v.s. beta`, finding `best beta`, if not an idea valley parabola, just pick the lowest wer from scanning points. (in fact, most of time, beta is not valley parabola curve, and wer hardly changed)
  * "FINE" scan `alpha` linespace in center = `last best alpha` width = `(--finetune_lm_alpha_max - --finetune_lm_alpha_min) / 2.0` (`beta` = `best beta`)
  * fitting valley parabola curve  on `wer v.s. alpha`, finding `best alpha`, if the fitting is not an idea valley parabola, just pick the lowest wer from scanning points.
  * use `best alpha` & `best beta` to do test() again, then compare with original one
  * export a report json file to `--finetune_output_file`

This finetune option could cost 3.x consuming of test epoch (`original test` + `finetune test` + `scanning` )

usage:
```
python -u DeepSpeech.py --test_files data/test.csv \
    --finetune_lm_csv_files data/test.csv \
    --finetune_lm_sampling_size 256 \
    --finetune_output_file data/lm/finetune.json
```
output report, I use the official model to test on voxforge test data, to generate the report, you must specify the `--test_files` and `--finetune_output_file`:
```
{
    "origin_test_result": {
        "alpha": 0.75,
        "beta": 1.85,
        "samples_wer": 0.19957946381636588,
        "samples_cer": 0.1047226538747382
    },
    "finetune_test_result": {
        "alpha": 0.9464133991551923,
        "beta": 2.0,
        "samples_wer": 0.19852812335728054,
        "samples_cer": 0.10810080399972975
    },
    "finetune_parameters": {
        "sample_size": 256,
        "alpha_min": 0.5,
        "alpha_max": 1.5,
        "alpha_steps": 7,
        "beta_min": 1.0,
        "beta_max": 2.0,
        "beta_steps": 5,
        "csv_files": "data/test.csv"
    }
}
```

### Other modification
----
* fix evaluate.py missing `import sys`
* return samples from test() for comparing results
* remove `test_csvs = FLAGS.test_files.split(',')` in `evaluate(test_csvs, create_model, try_loading)` function, this line overwrite the input argument `test_csvs`

reuben commented on util/FINETUNE_LM_PARAMETERS.md:
```diff
@@ -0,0 +1,43 @@
+===
+- Assume testing one sampling audio with ![](https://latex.codecogs.com/gif.latex?\\alpha_n) by `test()` function can get ![](https://latex.codecogs.com/gif.latex?output_n), where `output` can be considered as `wer`/`cer`/`word_distance`/`char_distance`
+- Then if do an experiment on testing sampling data, with different `alpha` settings, we should get a bunch outputs: ![](https://latex.codecogs.com/gif.latex?\\textbf{Y}=\begin{bmatrix}output_0&output_1&\.\.\.&output_n\end{bmatrix})
+- To solve the best parameter issue by experiment data, we should introduce statistics method
```
What statistics method? Sampling? Please fully describe what you're hinting at here.

reuben commented on DeepSpeech.py:
```diff
@@ -960,6 +1027,54 @@ def main(_):
+                    'csv_files': FLAGS.finetune_lm_csv_files,
+                }
+            }
+            open(FLAGS.finetune_output_file, 'w+', encoding='utf8').write(json.dumps(report, indent=4))
```
Use open mode `'w'` here, not `'w+'`.

reuben commented on DeepSpeech.py:
```diff
@@ -670,12 +671,77 @@ def __call__(self, progress, data, **kwargs):
         log_info('FINISHED optimization in {}'.format(datetime.utcnow() - train_start_time))
     log_debug('Session closed.')
 
+def finetune_lm_params():
```
Please move this into a separate file, similar to how `evaluate.py` is structured.

reuben commented on DeepSpeech.py:
```diff
@@ -670,12 +671,77 @@ def __call__(self, progress, data, **kwargs):
+    tmp_csv_path = os.path.join(FLAGS.finetune_temp_dir, '__TMP_FINETUNE_SAMPLE__.csv')
+    df = read_csvs(FLAGS.finetune_lm_csv_files.split(','))
+    n_sample = min(df.shape[0], FLAGS.finetune_lm_sampling_size)
+    df = df.sample(replace=False, n=n_sample)
```
`DataFrame.sample`'s `replace` parameter is already False by default.

reuben commented on DeepSpeech.py:
```diff
@@ -670,12 +671,77 @@ def __call__(self, progress, data, **kwargs):
 
+def finetune_lm_params():
+    original_alpha, original_beta = FLAGS.lm_alpha, FLAGS.lm_beta
+    tmp_csv_path = os.path.join(FLAGS.finetune_temp_dir, '__TMP_FINETUNE_SAMPLE__.csv')
```
Use the `tempfile` module instead of the temp dir flag and hardcoding a file name.

reuben commented on DeepSpeech.py:
```diff
@@ -670,12 +671,77 @@ def __call__(self, progress, data, **kwargs):
+    original_alpha, original_beta = FLAGS.lm_alpha, FLAGS.lm_beta
+    tmp_csv_path = os.path.join(FLAGS.finetune_temp_dir, '__TMP_FINETUNE_SAMPLE__.csv')
+    df = read_csvs(FLAGS.finetune_lm_csv_files.split(','))
+    n_sample = min(df.shape[0], FLAGS.finetune_lm_sampling_size)
```
nit: `len(df)` is more readable than `df.shape[0]`

reuben commented on util/finetune_lm_params.py:
```diff
@@ -0,0 +1,84 @@
+    return max(min(- c1 / 2.0 / c0, param_max), param_min)
+
+
+def finetune_linear(param_result_list, param_min, param_max):
```
Looks like this function is unused? Please remove it.

reuben commented on util/finetune_lm_params.py:
```diff
@@ -0,0 +1,84 @@
+        return pick_lowest_param(param_result_list)
+
+    # the parabola has minimum param
+    return max(min(- c1 / 2.0 / c0, param_max), param_min)
```
Where did the `2.0` term here come from?

reuben commented on util/FINETUNE_LM_PARAMETERS.md:
```diff
@@ -0,0 +1,43 @@
+Solve Best LM Paramters
+===
+- Now, we have `pseudo inverse tool` to fit every parabola curve, the rest job is coding then
+- About how to scan and sampling the data is not such absolutely
```
What?

reuben replied:
This entire section doesn't add much. Can you either expand it or remove it please?

reuben commented on DeepSpeech.py:
```diff
@@ -960,6 +1027,54 @@ def main(_):
         tfv1.reset_default_graph()
         do_single_file_inference(FLAGS.one_shot_infer)
 
+    if FLAGS.finetune_lm_csv_files:
```
Please move this into a separate file, similar to how `evaluate.py` is structured.

reuben commented on DeepSpeech.py:
```diff
@@ -960,6 +1027,54 @@ def main(_):
+        ))
+        if FLAGS.test_files:
+            FLAGS.lm_alpha = best_alpha
+            FLAGS.lm_beta = best_beta
```
Instead of overwriting the global flag, refactor the called functions to be able to take these values as parameters and pass them in.

reuben commented on DeepSpeech.py:
```diff
@@ -960,6 +1027,54 @@ def main(_):
 
+    if FLAGS.finetune_lm_csv_files:
+        original_report_count = FLAGS.report_count
+        FLAGS.report_count = 0
```
Instead of overwriting the global flag, refactor the called functions to be able to take these values as parameters and pass them in.

reuben replied:
And since this is just bypassing the report file generation done by `test`, you could just call `evaluate` directly, as that's all `test` does.

reuben commented on DeepSpeech.py:
```diff
@@ -670,12 +671,77 @@ def __call__(self, progress, data, **kwargs):
+
+    best_alpha = finetune_parabola(stats, FLAGS.finetune_lm_alpha_min, FLAGS.finetune_lm_alpha_max)
+    print('[FINETUNE LM] Set best alpha = {}'.format(best_alpha))
+    FLAGS.lm_alpha = best_alpha
```
Don't overwrite the global flag here.

reuben commented on DeepSpeech.py:
```diff
@@ -670,12 +671,77 @@ def __call__(self, progress, data, **kwargs):
+    # best_beta = pick_lowest_param(stats)
+    best_beta = finetune_parabola(stats, FLAGS.finetune_lm_beta_min, FLAGS.finetune_lm_beta_max)
+    print('[FINETUNE LM] Set best beta = {}'.format(best_beta))
+    FLAGS.lm_beta = best_beta
```
Instead of overwriting the global flag, refactor the called functions to be able to take these values as parameters and pass them in.

reuben commented on DeepSpeech.py:
```diff
@@ -670,12 +671,77 @@ def __call__(self, progress, data, **kwargs):
+    alpha_width = (FLAGS.finetune_lm_alpha_max - FLAGS.finetune_lm_alpha_min) / 4.0
+    for alpha in np.linspace(start=best_alpha - alpha_width, stop=best_alpha + alpha_width, num=FLAGS.finetune_lm_alpha_steps):
+        print('\t>>>> [FINETUNE LM] Test alpha = {} <<<<'.format(alpha))
+        FLAGS.lm_alpha = alpha
```
Instead of overwriting the global flag, refactor the called functions to be able to take these values as parameters and pass them in.

reuben commented on DeepSpeech.py:
```diff
@@ -670,12 +671,77 @@ def __call__(self, progress, data, **kwargs):
+    stats = []
+    for beta in np.linspace(start=FLAGS.finetune_lm_beta_min, stop=FLAGS.finetune_lm_beta_max, num=FLAGS.finetune_lm_beta_steps):
+        print('\t>>>> [FINETUNE LM] Test beta = {} <<<<'.format(beta))
+        FLAGS.lm_beta = beta
```
Instead of overwriting the global flag, refactor the called functions to be able to take these values as parameters and pass them in.

reuben commented on DeepSpeech.py:
```diff
@@ -670,12 +671,77 @@ def __call__(self, progress, data, **kwargs):
+
+    best_alpha = finetune_parabola(stats, FLAGS.finetune_lm_alpha_min, FLAGS.finetune_lm_alpha_max)
+    print('[FINETUNE LM] Set best alpha = {}'.format(best_alpha))
+    FLAGS.lm_alpha = best_alpha
```
Instead of overwriting the global flag, refactor the called functions to be able to take these values as parameters and pass them in.

reuben commented on DeepSpeech.py:
```diff
@@ -670,12 +671,77 @@ def __call__(self, progress, data, **kwargs):
+    stats = []
+    for alpha in np.linspace(start=FLAGS.finetune_lm_alpha_min, stop=FLAGS.finetune_lm_alpha_max, num=FLAGS.finetune_lm_alpha_steps):
+        print('\t>>>> [FINETUNE LM] Test alpha = {} <<<<'.format(alpha))
+        FLAGS.lm_alpha = alpha
```
Instead of overwriting the global flag, refactor the called functions to be able to take these values as parameters and pass them in.

mychiux413 commented:
> Thanks for the PR! This is really cool. I made a review pass and have a few questions and comments about the code.
> 
> Did you run it with any of our pre-trained English models? I'm curious about the results you got.

I used official pretrained model to test `librispeech-test-clean.csv` (2xxx rows), with 512 sampling size, and the result was:
```
{
    "origin_test_result": {
        "alpha": 0.75,
        "beta": 1.85,
        "samples_wer": 0.07449196826011226,
        "samples_cer": 0.029717934533042625
    },
    "finetune_test_result": {
        "alpha": 0.8936087708980639,
        "beta": 1.0,
        "samples_wer": 0.07261467002128895,
        "samples_cer": 0.030412160525297667
    },
    "finetune_parameters": {
        "sample_size": 512,
        "alpha_min": 0.5,
        "alpha_max": 1.5,
        "alpha_steps": 7,
        "beta_min": 1.0,
        "beta_max": 2.0,
        "beta_steps": 5,
        "csv_files": "data/libri-test.csv"
    }
}
```
As we can see, the original parameters `0.75` `1.85` is already good enough, and the `cer` is even better than `finetune_test_result`, and my `tuning target` is `wer`.

I don't know how to plot here, but I can print raw data in scanning log:
```
alpha | wer
0.5   | 0.082927
0.66  | 0.078146
0.833 | 0.076488
1.0   | 0.075707
1.166 | 0.076000
1.33  | 0.079512
1.5   | 0.087805
```

To determine `how parabolic` is the curve might not have a absolute definition, it should depend on fitting `residual value`, but how small is good? I don’t have much sense right now.

kdavis-mozilla commented on util/FINETUNE_LM_PARAMETERS.md:
```diff
@@ -0,0 +1,43 @@
+This topic is to explain how did `finetune_lm_params.py` fit `alpha` and `beta`
+
+## Inspiration
+According to mesh plot result of [Paddle/DeepSpeech-Grid-Search](https://github.com/PaddlePaddle/DeepSpeech/raw/develop/docs/images/tuning_error_surface.png), I assume `BatchWER v.s. Alpha` is `ideal valley parabolic curve`, i.e. which fits this formula below:
```
This will be true near the minimum of the BatchWER as a function of Alpha graph (the first derivative vanishes and higher derivatives are dominated by the second order derivative), but away from the minimum I see no reason for the quadratic approximation to hold in general, though it may.

kdavis-mozilla commented on util/FINETUNE_LM_PARAMETERS.md:
```diff
@@ -0,0 +1,43 @@
+# Finetune Language Model Parameters: Alpha, Beta
```
General question: Should we use latex.codecogs.com? I find their renderings to be subpar.

In the past I've used [readme2tex](https://github.com/leegao/readme2tex) which gives much better results., e.g. see their [README.md](https://github.com/leegao/readme2tex/blob/master/README.md).

kdavis-mozilla commented on util/FINETUNE_LM_PARAMETERS.md:
```diff
@@ -0,0 +1,43 @@
+- Let's tidy the `(A.2)` coefficients as below:
+- Assume there exists a `lowest WER` satisfies the formula: ![B.1](https://latex.codecogs.com/gif.latex?WER_{Batch}=B_0\\alpha^2+B_1\\alpha+B_2 "B.1"),  `(B.1)`
+- Where ![](https://latex.codecogs.com/gif.latex?B_0>0,B_1<0,B_2>0)
+- Compare to `(A.1)`: ![](https://latex.codecogs.com/gif.latex?B_0=A_0) and ![](https://latex.codecogs.com/gif.latex?B_1=-2A_0\\alpha_{best})
```
@mychiux413 You don't really want to say here that B_0=A_0. You want to say that B_0 has an analogous role to A_0 in the alpha quadratic equation.

Similarly you don't want to say B_1 = -2A_0\\alpha_{best}. You want to say that B_1 has an analogous role to -2A_0\\alpha_{best} in the alpha quadratic equation.

kdavis-mozilla commented on util/FINETUNE_LM_PARAMETERS.md:
```diff
@@ -0,0 +1,43 @@
+- To solve the `best alpha`, I decompose the formula, and change the coefficient as below:
+- ![A.2](https://latex.codecogs.com/gif.latex?WER_{batch}=A_0\\alpha^2-2A_0\\alpha_{best}\\alpha+A_1\\alpha_{best}^2+A_1 "A.2"),  `(A.2)`
+
+Polynomial Idea Valley Parabola Formula
```
I don't understand the need for this section. All the results here can be derived directly and more simply from the section above without introducing the B's at all.

kdavis-mozilla commented on util/FINETUNE_LM_PARAMETERS.md:
```diff
@@ -0,0 +1,43 @@
+
+Valley Parabola Formula in Matrix Form
+===
+- Assume testing one sampling audio with ![](https://latex.codecogs.com/gif.latex?\\alpha_n) by `test()` function can get ![](https://latex.codecogs.com/gif.latex?output_n), where `output` can be considered as `wer`/`cer`/`word_distance`/`char_distance`
```
You should not use n to index alpha_n and output_n here. The n here is different from the n in the next line. Use i here or some other dummy index here to indicate the difference.

kdavis-mozilla commented on util/FINETUNE_LM_PARAMETERS.md:
```diff
@@ -0,0 +1,43 @@
+
+Solve Formula `(C.1)`
+===
+- Because we can never get absolute solution of `(B.1)`, unless iterate every float value on every single audio, so we use instead `statistics` to solve `(C.1)` closest to `(B.1)`
```
Could you clarify what you mean by "statistics" here?

kdavis-mozilla commented on util/FINETUNE_LM_PARAMETERS.md:
```diff
@@ -0,0 +1,43 @@
+Solve Formula `(C.1)`
+===
+- Because we can never get absolute solution of `(B.1)`, unless iterate every float value on every single audio, so we use instead `statistics` to solve `(C.1)` closest to `(B.1)`
+- I use `Pseudo Inverse` directly to solve ![](https://latex.codecogs.com/gif.latex?\\hat{\\textbf{C}^T}=\\textbf{X}^\\dagger\\hat{\\textbf{Y}}), in numpy, the `Pseudo Inverse` is `numpy.linalg.pinv`
```
Elsewhere "we" is used. Here "I" is used. I'd prefer to stick with "we", or "one", or... but not "I"

kdavis-mozilla commented:
General comment. This PR largely assumes that the parameters alpha and beta do not interact, i.e. one can minimize the WER by varying alpha then fix alpha and minimize the WER by varying beta. I doubt that alpha and beta do not interact. Is there any justification for this?

kdavis-mozilla commented:
General comment: A simpler alternative to this that's easier to program and converges faster for a give computational budget is random search. See for example [Random search for hyper-parameter optimization](http://www.jmlr.org/papers/v13/bergstra12a.html). Why not use random search?

In addition random search does not make any of the "heavy handed" assumptions that are being made here, e.g. that WER is quadratic in alpha.

mychiux413 commented:
Sorry for the document, I'm not familiar with latex in markdown and also not familiar with some math terms in English, I rewrote a new document, but I haven't studied the `readme2tex` yet.

I also added a `random_search` algorithm,
if there are any better algorithms, we can add them into `finetune_lm_params.py`

After some tests on `random_search`, my observation is:
- `parabola` can fit faster than `random_search` in wider scanning range (0.5 < `alpha` < 1.5).
- `random_search` can find better(lower loss) position in narrow scanning range (set `scan radius` = 0.05).

So, my suggestion is:
For a new language model, the better policy is to scan `parabola` to find `best alpha/beta` then use it as initial position for `random_search`.
For a known language model, we can just scan `random_search` from an known position like pre-trained model use: (0.75, 1.85).

So I add a flag to switch tuning policy:
`--finetune_lm_method` = ['parabola', 'random_search', 'parabola+random_search']

lissyx commented:
@mychiux413 Does that PR still make sense now that we have the LM optimizer code ?

mychiux413 commented:
@lissyx  I don't actually have more resources to benchmark the performance between them, but we can close the PR, I think the two should not be too different, and depending on `optuna` should be more stable.
